### PR TITLE
Better logging for failures in qemu and host targets

### PR DESCRIPTION
**Pull request by hnj2**
If we introduce a deliberate error into out program:

```diff
diff --git a/victims/schoolbook32/main.c b/victims/schoolbook32/main.c
index a9d1ded..ac9f49f 100644
--- a/victims/schoolbook32/main.c
+++ b/victims/schoolbook32/main.c
@@ -53,6 +53,7 @@ static uint8_t recv_plain(int size, uint16_t* data)
 
     /* Sending output array */
     simpleserial_put('z', 1, (uint8_t *) &status);
+    *(int*)0 = 0;
     simpleserial_put('r', 2*result_length, (uint8_t *)result);
 
   return 0x00;

```
 we now have multiple possibilities to debug ...

### When a process or file closes unexpectedly an appropriate error is thrown:
```
(venv) root@failbox: ~/m4sc$ python -m m4sc test
...
FAILED vic... - m4sc.connections.VictimConnectionError: Process stopped unexpectedly!
FAILED vic... - m4sc.connections.VictimConnectionError: Input file unexpectedly at EOF!
```

### Remaining stdout and stderr, aswell as return codes of a process are logged now.
```
(venv) root@failbox: ~/m4sc$ python -m m4sc test
...
------------------------------------- Captured log teardown -------------------------------------
ERROR    m4sc.connections:connections.py:245 Process exited with return code -11 (SIGSEGV: segmentation fault due to illegal access of a memory segment)!
DEBUG    m4sc.connections:connections.py:250 ### Process stdout ###

DEBUG    m4sc.connections:connections.py:251 ### Process stderr ###
=================================================================================================
```

### New **host-gdb** target runs the application in gdb to give better logging.
```

(venv) root@failbox: ~/m4sc$ python -m m4sc test --target host-gdb
...
------------------------------------- Captured log teardown -------------------------------------
DEBUG    m4sc.connections:connections.py:224 Process exited normally!
DEBUG    m4sc.connections:connections.py:250 ### Process stdout ###

Program received signal SIGSEGV, Segmentation fault.
0x0000555555554ba3 in recv_plain (size=8, data=0x7fffffffc950) at victims/schoolbook32/main.c:56
56          *(int*)0 = 0;

DEBUG    m4sc.connections:connections.py:251 ### Process stderr ###
=================================================================================================
```

### CODE REVIEW

**Comment by hnj2** [comment]
This shold close #16

**Comment by nils-wisiol** [comment]
Nice! Very useful!

Three suggestions:

1. make `process_exit` a member method of `StreamConnection`
1. reduce code duplication in `create_qemu_connection`, `create_host_connection`, `create_host_gdb_connection`
1. replace `host-gdb` target by a `--debug` or `--debugger` flag that can be optionally added

In any event, we can merge this soon! :rocket:

**Comment by hnj2** [comment]
### Hew ho thanks for the suggestions!
>     1. make `process_exit` a member method of `StreamConnection`
>     2. reduce code duplication in `create_qemu_connection`, `create_host_connection`, `create_host_gdb_connection`
It'll take a bit of context_manager/python-oop digging ... I can't see a nice implementation, but I'll certainly try it :P

>     3. replace `host-gdb` target by a `--debug` or `--debugger` flag that can be optionally added
So I've added the `--debug` flag to the test command in the m4sc module ... but translated to the `-s` flag in pytest it's way too much information (try `python -m m4sc test --target host-gdb --debug`).

I think we probably want to try to start a gdb session for all targets anyway at some point?
So I'll go for the `--debugger` option.

**Comment by hnj2** [comment]
Hey ho @nils-wisiol I would like your feedback on the oop.
Notice that the argument parsing could be nicer, see #30 for further steps to clean it up.

File: m4sc/connections.py [hunk 1]
```diff
@@ -330,14 +332,25 @@ def __init__(self) -> None:
...
+        )
+        self.exit_stack.enter_context(
+            super().__init__(process, fin, fout)
+        )
+
+        # We need to close the files before trying to close the process.
+        # Some issues are cause by programs waiting ofr input otherwise.
+        self.exit_stack.callback(self.stdout.close)
+        self.exit_stack.callback(self.stdin.close)
+
```
Comment (hnj2) [inline m4sc/connections.py:349]: https://github.com/nils-wisiol/m4sc/pull/27#pullrequestreview-451363913
Well even though this is the pythonic way I would was it's actually more complicated :P

Comment (hnj2) [reply m4sc/connections.py:353]: I think i got the bloat minimized now

File: m4sc/__main__.py [hunk 2]
```diff
@@ -182,6 +182,6 @@ def _build_victim(self, victim: Victim, target: str, clean: bool):
 
 
 if __name__ == "__main__":
-    logging.basicConfig(level=logging.WARNING)
+    logging.basicConfig(level=logging.INFO)
```
Comment (hnj2) [inline m4sc/__main__.py:185]: Error reporting in during tests was broken.
We might want to use multiple loggers?
One for testing and one for the rest of the application?
@nils-wisiol

Comment (nils-wisiol) [reply m4sc/__main__.py:185]: Eventually, we should connect this to a command line parameter. For now, I'm fine with whatever works

**Comment by nils-wisiol** [comment]
lgtm!

**Pull request merged**